## Bag of words Model

A simplifying representation of text used in natural language processing and information retrieval. A text (sentence or a document) is represented as the bag of its words, disregarding grammar and word order. The multiplicity of the words are being kept. 


#### Pros and Cons of Bag of words model
- Pros
    - If the dataset is small and context is domain specific, using BoW model may work better than word embedding, because it's hard to find corresponding vector from pre-trained word embedding models
    - It's relative easy to build a baseline model. For example using scikit-learn library, just a few lines of code the model can be built. I will provide some examples below. 
    
- Cons
    - Large feature dimension considering each token is a feature
    - Sparcity: a dataset comprises mostly zero values

### Feature Extraction vs. Feature Selection
```Feature Extraction```: Transforming arbitary data, such as text or images into ```numerical``` features usable for machine learning


```Feature Selection```: A machine learning technique applied on these features


### Feature Extraction most intivitive ways:
    1. Assign a fixed integer id to each word occuring in any document of the training set (build a dictionary from words to integer indices)
    2. For each document #i, count the number of occurrences of each word w, and store it in x[i, j] as the value of feature #j where j is the index of word w in the dictionary

### high-dimensional ```sparse``` dataset

- A sparse dataset is a dataset that is comprised of mostly zero values
- Sparse matrices are distinct from matrices with mostly non-zero values, which are referred to as dense matrices *
- Sparse Matrices can cause problmes regards to space and time complexity
- Sparsity can be computed as below:
    - $sparsity = 1.0 - count_nonzero(A) / A.size$

### Examples for Sparse matrics for specific machine learning problems:
- Whether or not a user has watched a movie in a movie catalog
- Count of the number of words in a document corpus
- Whether or not a user has purchased a product in a product list


In [101]:
# dense to sparse
from numpy import array
from scipy.sparse import csr_matrix
from numpy import count_nonzero

# create dense matrix
A = array([[1, 0, 0, 1, 0, 0], [0, 0, 2, 0, 0, 1], [0, 0, 0, 2, 0, 0]])
print('''The input array: 
{}
'''.format(A))
print()
# convert to sparse matrix (CSR method)
S = csr_matrix(A)
print('''Converted to Sparse Matrix: 
{}
'''.format(S))
print()
# reconstruct dense matrix
B = S.todense()
print('''Convert back to Dense Matrix: 
{}
'''.format(B))
print()

sparsity = 1.0 - count_nonzero(A) / A.size
print('''
The array size is: {}
The input array sparsity: {}'''.format(A.size, sparsity))


The input array: 
[[1 0 0 1 0 0]
 [0 0 2 0 0 1]
 [0 0 0 2 0 0]]


Converted to Sparse Matrix: 
  (0, 0)	1
  (0, 3)	1
  (1, 2)	2
  (1, 5)	1
  (2, 3)	2


Convert back to Dense Matrix: 
[[1 0 0 1 0 0]
 [0 0 2 0 0 1]
 [0 0 0 2 0 0]]



The array size is: 18
The input array sparsity: 0.7222222222222222


### Feature Extraction -  Word Counting
- ```tokenization```: strings and int id for each possible token. Potentially remove the stop words (the most common words in a language)
- ```counting```: the occurrences of tokens in each document


In [104]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction import stop_words

# two sentences with 21 words, forms i=2, j = 21 matrix
text = ['' for i in range(2)]
text[0]="SciPy provides tools for creating sparse matrices using multiple data structures"
text[1]="A dense matrix stored in a NumPy array can be converted into a sparse matrix"
print('''

Here is the source text:
{}
'''.format(text))


# Create the transform
vectorizer = CountVectorizer()
#Tokenize and build vocab
vectorizer.fit(text)
#summarize the vocabulary
print('Below is the vocabulary based on the text above:')
v = vectorizer.vocabulary_
print(v)
print('')
print('Total vocabularies: {} '.format(len(v)))
print('')
print("token(word) 'sparse' has index: {} ".format(vectorizer.vocabulary_.get('sparse')))
print()
print("index 11 has token (word): '{}' ".format(vectorizer.get_feature_names()[11]))
print()

print("Let's count the the tokens")
vector = vectorizer.transform(text)
print('''After transform() to count the tokens, the word count feature has the following shape: 
{}'''.format(vector.shape))

print('''
The word count feature type is: 
{}'''.format(type(vector)))

print('''
Word count feature presented as sparse matrix:
{}'''.format(vector))

va = vector.toarray()
print('''
Convert word count feature from sparse matrix to dense matrix(2 by 21):
{}'''.format(va))

sparsity = 1.0 - count_nonzero(va) / va.size
print('The input array sparsity: {}'.format(sparsity))


print('''
The stop words list in skit learn library:
{}
'''.format(stop_words.ENGLISH_STOP_WORDS))




Here is the source text:
['SciPy provides tools for creating sparse matrices using multiple data structures', 'A dense matrix stored in a NumPy array can be converted into a sparse matrix']

Below is the vocabulary based on the text above:
{'scipy': 15, 'provides': 14, 'tools': 19, 'for': 7, 'creating': 4, 'sparse': 16, 'matrices': 10, 'using': 20, 'multiple': 12, 'data': 5, 'structures': 18, 'dense': 6, 'matrix': 11, 'stored': 17, 'in': 8, 'numpy': 13, 'array': 0, 'can': 2, 'be': 1, 'converted': 3, 'into': 9}

Total vocabularies (CountVectorizer removes the stopwords): 21 

token(word) 'sparse' has index: 16 

index 11 has token (word): 'matrix' 

Let's count the the tokens
After transform() to count the tokens, the word count feature has the following shape: 
(2, 21)

The word count feature type is: 
<class 'scipy.sparse.csr.csr_matrix'>

Word count feature presented as sparse matrix:
  (0, 4)	1
  (0, 5)	1
  (0, 7)	1
  (0, 10)	1
  (0, 12)	1
  (0, 14)	1
  (0, 15)	1
  (0, 16)	1
  (0,

### Feature Extraction - TFIDF


- Word counts are good starting point
- Issue: some words will appear many times, and their large counts will not be very meaningful in the encoded matrix
- An alternative way is to calculate word frequencies using TF-IDF
    - Term Frequence (TF): summarizes how often a given word appears within a document
    - Inverse Document Frequency (IDF): term weight function, and a measure of how much information the word provides
    
- Python libraries providing out of box TFIDF calcuation
    - scikit learn
    - Gensim 
    - PySpark ML lib
    - nltk
    

### TF-IDF Calculation (Term Frequency - Inverse Document Frequency):

### $tfidf(t,d) = tf(t,d) * idf(t)$
 
- n: the total number of documents in the document set 


-  df(t): * the number of documents in the document set that contain term t (smooth_idf = False):
### $idf(t) = log \frac{n}{df(t)} + 1$


- df(t): * tfidf with smooth_idf = True *
### $idf(t) = log \frac{1+n}{1+df(t)} + 1$


- Different flavored tfidf (standard textbook):
### $idf(t) = log \frac{n}{1+df(t)}$


- The resulting tf-idf vectors are then ```normalized``` by Euclidean norm (a function that assigns a strictly positive length or size to each vector in a vector space)
## $v_{norm} = \frac{V}{|v|} = \frac{[v_1, v_2, ..., v_n]}{\sqrt{v_1^2 + v_2^2 + ... + v_n^2}}$


In [112]:
counts = [
[3, 0, 1],
[2, 0, 0],
[3, 0, 0],
[4, 0, 0],
[3, 2, 0],
[3, 0, 2]
]

counts

[[3, 0, 1], [2, 0, 0], [3, 0, 0], [4, 0, 0], [3, 2, 0], [3, 0, 2]]

In [113]:
tfidf = transformer.fit_transform(counts)
tfidf.toarray()


array([[0.81940995, 0.        , 0.57320793],
       [1.        , 0.        , 0.        ],
       [1.        , 0.        , 0.        ],
       [1.        , 0.        , 0.        ],
       [0.47330339, 0.88089948, 0.        ],
       [0.58149261, 0.        , 0.81355169]])

#### Steps of calcuation:



- doc1 term1 calcuation (Smooth = false): 
    - n = 6 (6 documents)
    - $df(t)_{term1} = 6$ (term1 appears in all 6 documents)
    - $idf(t)_{term1} = log\frac{n}{df(t)} + 1 = log(\frac {6}{6}) + 1 = 1$
    - $tfidf_{term1} = tf * idf = 3 * 1 = 3$ (term1 appears 3 times in document 1)



- doc1 term2 calculation:
    - n = 6
    - $df(t)_{term2} = 1$ (term2 appears in document No 5)
    - $idf(t)_{term2} = log\frac{n}{df(t)} + 1 = log(\frac {6}{1}) + 1 \approx 2.792$
    - $tfidf_{term2} = tf * idf \approx 0 * 2.792 = 0$ (term1 appears 3 times in document 1)
   
   

- doc1 term3 calculation:
    - n = 6
    - $df(t)_{term3} = 2$ (term3 appears in document No 1 and document No 6)
    - $idf(t)_{term3} = log\frac{n}{df(t)} + 1 = log(\frac{6}{2}) + 1 \approx 2.0986$
    - $tfidf_{term3} = tf * idf = 1 * 2.0986 \approx 2.0986$ (term1 appears 3 times in document 1)
    



- Use Euclidean norm, tfidf for first document:
    - $\frac{[3,  0,  2.0986]}{\sqrt{3^2 + 0^2 + 2.0986^2}} = [0.819, 0, 0.573]$
    

In [114]:
transform = TfidfTransformer(smooth_idf=True)
tfidf = transform.fit_transform(counts)
tfidf.toarray()

array([[0.85151335, 0.        , 0.52433293],
       [1.        , 0.        , 0.        ],
       [1.        , 0.        , 0.        ],
       [1.        , 0.        , 0.        ],
       [0.55422893, 0.83236428, 0.        ],
       [0.63035731, 0.        , 0.77630514]])

#### Steps of calcuation using Smooth = True:



- doc1 term1 calcuation (Smooth = True): 
    - n = 6 (6 documents)
    - $df(t)_{term1} = 6$ (term1 appears in all 6 documents)
    - $idf(t)_{term1} = log\frac{n + 1}{df(t) + 1} + 1 = log(\frac{7}{7}) + 1 = 0$
    - $tfidf_{term1} = tf * idf = 3 * 1 = 3$ (term1 appears 3 times in document 1)



- doc1 term2 calculation:
    - n = 6
    - $df(t)_{term2} = 1$ (term2 appears in document No 5)
    - $idf(t)_{term2} = log\frac{n + 1}{df(t) + 1} + 1 = log(\frac{7}{2}) + 1 \approx 2.2527$
    - $tfidf_{term2} = tf * idf = 0 * 2.2527 = 0$ (term1 appears 3 times in document 1)
   
   

- doc1 term3 calculation:
    - n = 6
    - $df(t)_{term3} = 2$ (term3 appears in document No 1 and document No 6)
    - $idf(t)_{term3} = log\frac{n + 1}{df(t) + 1} + 1 = log(\frac{7}{3}) + 1 \approx 1.8473$
    - $tfidf_{term3} = tf * idf = 1 * 1.8473 \approx 1.8473$ (term1 appears 3 times in document 1)
    



- Use Euclidean norm, tfidf for first document:
    - $\frac{[3, 0, 1.8473]}{\sqrt{3^2 + 0^2 + 1.8473^2}} = [0.8515, 0, 0.5243]$

In [92]:
from sklearn.feature_extraction.text import TfidfTransformer

transformer = TfidfTransformer(smooth_idf=False)
print('''Transformer parameters in sklearn: 
{} '''.format(transformer))

print('''

In the previous word count example we got sparse matrix of 'vector' for two sentences in 'text'
This sparse matrix will be used to Tfidf transformation ''')
tfidf = transformer.fit(vector)

print('''

Let's transform into Tfidf matrix
''')
tfidf =transformer.transform(vector)

print('''
Tfidf sparse matrix output:
{}
'''.format(tfidf))

print('''
Tfidf sparse matrix shape:
{}
'''.format(tfidf.shape))

print('''
Convert tfidf into dense matrix:
{}
'''.format(tfidf.toarray()))

idx = vectorizer.vocabulary_.get('sparse')
tfidf_idx_1 = tfidf[0,idx]
tfidf_idx_2 = tfidf[1,idx]
print('''

tfidf for token 'sparse' feature index: {}
tfidf for token 'sparse' in document 1: {}
tfidf for token 'sparse' in document 2: {}
'''.format(idx, tfidf_idx_1, tfidf_idx_2))


idx = vectorizer.vocabulary_.get('data')
tfidf_idx_1 = tfidf[0,idx]
tfidf_idx_2 = tfidf[1,idx]
print('''

tfidf for token 'data' feature index: {}
tfidf for token 'data' in document 1: {}
tfidf for token 'data' in document 2: {}
'''.format(idx, tfidf_idx_1, tfidf_idx_2))


Transformer parameters in sklearn: 
TfidfTransformer(norm='l2', smooth_idf=False, sublinear_tf=False,
         use_idf=True) 


In the previous word count example we got sparse matrix of 'vector' for two sentences in 'text'
This sparse matrix will be used to Tfidf transformation 


Let's transform into Tfidf matrix


Tfidf sparse matrix output:
  (0, 20)	0.3108525457240639
  (0, 19)	0.3108525457240639
  (0, 18)	0.3108525457240639
  (0, 16)	0.18359452107480756
  (0, 15)	0.3108525457240639
  (0, 14)	0.3108525457240639
  (0, 12)	0.3108525457240639
  (0, 10)	0.3108525457240639
  (0, 7)	0.3108525457240639
  (0, 5)	0.3108525457240639
  (0, 4)	0.3108525457240639
  (1, 17)	0.27370229648379657
  (1, 16)	0.16165298541458145
  (1, 13)	0.27370229648379657
  (1, 11)	0.5474045929675931
  (1, 9)	0.27370229648379657
  (1, 8)	0.27370229648379657
  (1, 6)	0.27370229648379657
  (1, 3)	0.27370229648379657
  (1, 2)	0.27370229648379657
  (1, 1)	0.27370229648379657
  (1, 0)	0.27370229648379657


Tfidf sparse

### Next Step: Build model Using Extracted Features

i.e. Text Classification
- LDA
- Naive Bayes
- Decision Tree
...

### References:
- [working with text data] https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html
- [Feature Extraction] https://scikit-learn.org/stable/modules/feature_extraction.html#
- [Feature Selection] https://scikit-learn.org/stable/modules/feature_selection.html#feature-selection
- [Sparse Matrix and Machine Learning] https://machinelearningmastery.com/sparse-matrices-for-machine-learning/
- [Bag of Words Model] https://en.wikipedia.org/wiki/Bag-of-words_model
- [spark ML lib Feature Extraction] https://spark.apache.org/docs/latest/mllib-feature-extraction.html